source/server/admin/ungrouped_stats_request.cc

```diff
@@ -0,0 +1,102 @@
+#include "source/server/admin/ungrouped_stats_request.h"
+
+namespace Envoy {
+namespace Server {
+
+UngroupedStatsRequest::UngroupedStatsRequest(Stats::Store& stats, const StatsParams& params,
+                                             UrlHandlerFn url_handler_fn)
+    : StatsRequest(stats, params, url_handler_fn) {
+  phases_ = {Phase::TextReadouts, Phase::CountersAndGauges, Phase::Histograms};
+  switch (params_.type_) {
+  case StatsType::TextReadouts:
+  case StatsType::All:
+    phase_index_ = 0;
+    break;
+  case StatsType::Counters:
+  case StatsType::Gauges:
+    phase_index_ = 1;
+    break;
+  case StatsType::Histograms:
+    phase_index_ = 2;
+    break;
+  }
+}
+
+template <class StatType> Stats::IterateFn<StatType> UngroupedStatsRequest::saveMatchingStat() {
+  return [this](const Stats::RefcountPtr<StatType>& stat) -> bool {
+    // Check if used.
+    if (params_.used_only_ && !stat->used()) {
+      return true;
+    }
+
+    // Capture the name if we did not early-exit due to used_only -- we'll use
+    // the name for both filtering and for capturing the stat in the map.
+    // stat->name() takes a symbol table lock and builds a string, so we only
+    // want to call it once.
+    //
+    // This duplicates logic in saveMatchingStat in grouped_stats_request.cc, but
+    // differs in that Prometheus only uses stat->name() for filtering, not
+    // rendering, so it only grabs the name if there's a filter.
+    std::string name = stat->name();
+
+    // Check if filtered.
+    if (params_.re2_filter_ != nullptr && !re2::RE2::PartialMatch(name, *params_.re2_filter_)) {
+      return true;
+    }
+
+    stat_map_[name] = stat;
+    return true;
+  };
+}
+
+Stats::IterateFn<Stats::TextReadout> UngroupedStatsRequest::saveMatchingStatForTextReadout() {
+  return saveMatchingStat<Stats::TextReadout>();
+}
+
+Stats::IterateFn<Stats::Gauge> UngroupedStatsRequest::saveMatchingStatForGauge() {
+  return saveMatchingStat<Stats::Gauge>();
+}
+
+Stats::IterateFn<Stats::Counter> UngroupedStatsRequest::saveMatchingStatForCounter() {
+  return saveMatchingStat<Stats::Counter>();
+}
+
+Stats::IterateFn<Stats::Histogram> UngroupedStatsRequest::saveMatchingStatForHistogram() {
+  return saveMatchingStat<Stats::Histogram>();
+}
+
+template <class SharedStatType>
+void UngroupedStatsRequest::renderStat(const std::string& name, Buffer::Instance& response,
+                                       const StatOrScopes& variant) {
+  auto stat = absl::get<SharedStatType>(variant);
+  dynamic_cast<StatsRender*>(render_.get())->generate(response, name, stat->value());
```

Comment:
it would be good to brainstorm about how to get rid of the dynamic_cast. I hadn't noticed these before.

They are a mild anti-pattern IMO, and also they are slow.

They don't seem too hard to get rid of -- can't we have the `render_` object in UngroupedStatsRequest be a StatsRender?